python/packages/autogen-agentchat/src/autogen_agentchat/agents/_code_executor_agent.py

```diff
@@ -409,93 +421,130 @@ async def on_messages_stream(
+            if not code_blocks:
+                yield Response(
+                    chat_message=TextMessage(
+                        content=str(model_result.content),
+                        source=agent_name,
+                    )
+                )
+                return
+
+            # Step 7: Yield a CodeGenerationEvent
+            inferred_text_message: CodeGenerationEvent = CodeGenerationEvent(
+                retry_attempt=nth_try,
                 content=model_result.content,
+                code_blocks=code_blocks,
                 source=agent_name,
-                thought=getattr(model_result, "thought", None),
             )
-        )
 
-        code_blocks = self._extract_markdown_code_blocks(str(model_result.content))
+            yield inferred_text_message
 
-        if not code_blocks:
-            yield Response(
-                chat_message=TextMessage(
-                    content=str(model_result.content),
+            # Step 8: Execute the extracted code blocks
+            execution_result = await self.execute_code_block(inferred_text_message.code_blocks, cancellation_token)
+
+            # Step 9: Update model context with the code execution result
+            await model_context.add_message(
+                UserMessage(
+                    content=execution_result.output,
                     source=agent_name,
                 )
             )
-            return
 
-        # NOTE: error: Argument of type "str | List[FunctionCall]" cannot be assigned to parameter "content" of type "str" in function "__init__".
-        #       For now we can assume that there are no FunctionCalls in the response because we are not providing tools to the CodeExecutorAgent.
-        #       So, for now we cast model_result.content to string
-        inferred_text_message: CodeGenerationEvent = CodeGenerationEvent(
-            content=str(model_result.content),
-            code_blocks=code_blocks,
-            source=agent_name,
-        )
+            # Step 10: Yield a CodeExecutionEvent
+            yield CodeExecutionEvent(retry_attempt=nth_try, result=execution_result, source=self.name)
 
-        yield inferred_text_message
+            # If execution was successful or last retry, then exit
+            if execution_result.exit_code == 0 or nth_try == max_retries_on_error:
+                break
 
-        execution_result = await self.execute_code_block(inferred_text_message.code_blocks, cancellation_token)
+            # Step 11: If exit code is non-zero and retries are available then
+            #          make an inference asking if we should retry or not
+            chat_context = await model_context.get_messages()
 
-        # Add the code execution result to the model context
-        await model_context.add_message(
-            UserMessage(
-                content=execution_result.result.output,
-                source=agent_name,
+            retry_prompt = (
```

Comment:
We need to use class variable for this retry prompt. 

We don't need the error output I think, it is already included in the `model_context` and is part of the `chat_context`. 

Also, we need to update the prompt with some fallbacks:
1. Structured output when `model_client.model_info["structured_output"]`.
2. JSON mode when `model_client.model_info["json_output"]`, i.e., setting `json_output=True`
3. Best-effort JSON generation

In case 2 and 3, we need to explictly set the expected JSON format in the prompt itself.

Reply:
Actually, let's do that for a future PR.